### PR TITLE
@types/stripe: Update bank account-related types

### PR DESCRIPTION
**Pull request by jekh**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: 
https://stripe.com/docs/api/customer_bank_accounts
(See also PR comments inline)
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This PR updates several types related to bank accounts. I'll include inline comments with links to the stripe docs.

### CODE REVIEW

File: types/stripe/index.d.ts [hunk 1]
```diff
@@ -411,7 +411,7 @@ declare namespace Stripe {
                  * "individual" or "company". This field is required when attaching the
                  * bank account to a customer object.
                  */
-                account_holder_type?: string;
+                account_holder_type?: "individual" | "company" | null;
```
Comment (jekh) [inline types/stripe/index.d.ts]: https://stripe.com/docs/api/customer_bank_accounts/object#customer_bank_account_object-account_holder_type

Stripe will set this to `null` in responses if the value is not set.

File: types/stripe/index.d.ts [hunk 2]
```diff
@@ -902,7 +902,7 @@ declare namespace Stripe {
              * For most Stripe users, the source of every charge is a credit or debit card.
              * This hash is then the card object describing that card.
              */
-            source: cards.ICard | bitcoinReceivers.IBitcoinReceiver;
+            source: cards.ICard | bitcoinReceivers.IBitcoinReceiver | bankAccounts.IBankAccount;
```
Comment (jekh) [inline types/stripe/index.d.ts]: The docs mention credit or debit cards, but bank accounts (among others) are also valid: https://stripe.com/docs/api/charges/create#create_charge-source

File: types/stripe/index.d.ts [hunk 3]
```diff
@@ -1344,12 +1344,12 @@ declare namespace Stripe {
             /**
              * The currency the customer can be charged in for recurring billing purposes (subscriptions, invoices, invoice items).
              */
-            currency: string;
+            currency: string | null;
```
Comment (jekh) [inline types/stripe/index.d.ts]: Stripe sets this to `null` if the value is not set.

File: types/stripe/index.d.ts [hunk 4]
```diff
@@ -4073,12 +4073,12 @@ declare namespace Stripe {
 
     namespace bankAccounts {
         interface IBankAccount extends IBankAccountHash {
-            account: string;
+            account?: string;
```
Comment (jekh) [inline types/stripe/index.d.ts]: Stripe does not always return this value. There is no documentation on what this value is at https://stripe.com/docs/api/customer_bank_accounts/object#customer_bank_account_object-account

File: types/stripe/index.d.ts [hunk 5]
```diff
@@ -4073,12 +4073,12 @@ declare namespace Stripe {
...
 
             /**
              * This indicates whether or not this bank account is the default external account for its currency.
              */
-            default_for_currency: boolean;
+            default_for_currency?: boolean;
```
Comment (jekh) [inline types/stripe/index.d.ts]: This does not appear to be present in most IBankAccount responses, and it is not mentioned in the docs at https://stripe.com/docs/api/customer_bank_accounts/object

File: types/stripe/index.d.ts [hunk 6]
```diff
@@ -4090,20 +4090,20 @@ declare namespace Stripe {
 
         interface IBankAccountHash extends IResourceObject {
             /**
-             * value is "customer_bank_account"
+             * value is "bank_account"
              */
-            object: "customer_bank_account";
+            object: "bank_account";
```
Comment (jekh) [inline types/stripe/index.d.ts]: `bank_account` is the proper name of the object, and it is what stripe currently returns: https://stripe.com/docs/api/customer_bank_accounts/object#customer_bank_account_object-object

The only reference I could find to `customer_bank_account` is an old IRC chat that mentions it was a documentation mistake: https://webcache.googleusercontent.com/search?q=cache:RXkeQmg0xLUJ:https://botbot.me/freenode/stripe/2016-04-14/%3Fpage%3D5+&cd=3&hl=en&ct=clnk&gl=us

File: types/stripe/index.d.ts [hunk 7]
```diff
@@ -4090,20 +4090,20 @@ declare namespace Stripe {
...
 
             /**
              * The name of the person or business that owns the bank account.
              */
-            account_holder_name: string;
+            account_holder_name: string | null;
```
Comment (jekh) [inline types/stripe/index.d.ts]: Stripe returns `null` if this value is not set.

File: types/stripe/index.d.ts [hunk 8]
```diff
@@ -4090,20 +4090,20 @@ declare namespace Stripe {
...
 
             /**
              * The type of entity that holds the account. This can be either
              * "individual"" or "company".
              */
-            account_holder_type: string;
+            account_holder_type: "individual" | "company" | null;
```
Comment (jekh) [inline types/stripe/index.d.ts]: These are the only valid values: https://stripe.com/docs/api/customer_bank_accounts/object#customer_bank_account_object-account_holder_type

Stripe will return `null` if the value is not set.

File: types/stripe/index.d.ts [hunk 9]
```diff
@@ -4120,6 +4120,8 @@ declare namespace Stripe {
              */
             currency: string;
 
+            customer?: string;
```
Comment (jekh) [inline types/stripe/index.d.ts]: This field is mentioned in the stripe API docs without explanation. In practice, it does appear to be the customer ID.
https://stripe.com/docs/api/customer_bank_accounts/object#customer_bank_account_object-customer

File: types/stripe/index.d.ts [hunk 10]
```diff
@@ -4145,7 +4147,7 @@ declare namespace Stripe {
              * transfer sent to this bank account fails, we’ll set the status to "errored""
              * and will not continue to send transfers until the bank details are updated.
              */
-            status: "new" | "validated" | "verified" | "verification_failed" | "error" ;
+            status: "new" | "validated" | "verified" | "verification_failed" | "errored";
```
Comment (jekh) [inline types/stripe/index.d.ts]: Fixed "errored" value: https://stripe.com/docs/api/customer_bank_accounts/object#customer_bank_account_object-status

File: types/stripe/index.d.ts [hunk 11]
```diff
@@ -4194,12 +4196,19 @@ declare namespace Stripe {
...
+            account_holder_name?: string | null;
 
             /**
              * The type of entity that holds the account. This can be either "individual" or "company".
              */
-            account_holder_type?: "individual" | "company";
+            account_holder_type?: "individual" | "company" | null;
+        }
+
+        interface IBankAccountVerifyOptions {
```
Comment (jekh) [inline types/stripe/index.d.ts]: There was previously no type for the `verirfySource` parameters: https://stripe.com/docs/api/customer_bank_accounts/verify

File: types/stripe/index.d.ts [hunk 12]
```diff
@@ -4452,15 +4461,15 @@ declare namespace Stripe {
             /**
              * If address_line1 was provided, results of the check: pass, fail, unavailable, or unchecked.
              */
-            address_line1_check: string | null;
+            address_line1_check: "pass" | "fail" | "unavailable" | "unchecked" | null;
```
Comment (jekh) [inline types/stripe/index.d.ts]: Values: https://stripe.com/docs/api/cards/object#card_object-address_line1_check

File: types/stripe/index.d.ts [hunk 13]
```diff
@@ -4452,15 +4461,15 @@ declare namespace Stripe {
...
             address_line2: string | null;
             address_state: string | null;
-            address_zip: string;
+            address_zip: string | null;
```
Comment (jekh) [inline types/stripe/index.d.ts]: This will be `null` if no zip code was provided.

File: types/stripe/index.d.ts [hunk 14]
```diff
@@ -4452,15 +4461,15 @@ declare namespace Stripe {
...
 
             /**
              * If address_zip was provided, results of the check: pass, fail, unavailable, or unchecked.
              */
-            address_zip_check: "pass" | "fail" | "unavailable" | "unchecked" ;
+            address_zip_check: "pass" | "fail" | "unavailable" | "unchecked" | null;
```
Comment (jekh) [inline types/stripe/index.d.ts]: Values: https://stripe.com/docs/api/cards/object#card_object-address_zip_check

File: types/stripe/index.d.ts [hunk 15]
```diff
@@ -6271,8 +6280,18 @@ declare namespace Stripe {
...
+             * A customer's bank account must first be verified before it can be charged. Stripe supports instant verification using Plaid for many of
+             * the most popular banks. If your customer's bank is not supported or you do not wish to integrate with Plaid, you must manually verify
+             * the customer's bank account using the API.
+             *
+             * @returns Returns the bank account object.
+             *
+             * @param customerId The ID of the customer whose source needs to be verified.
+             * @param sourceId The ID of the source to be verified.
+             */
+            verifySource(customerId: string, sourceId: string, data: bankAccounts.IBankAccountVerifyOptions, options: HeaderOptions): Promise<bankAccounts.IBankAccount>;
```
Comment (jekh) [inline types/stripe/index.d.ts]: This previously did not include the proper types for the verification amounts: https://stripe.com/docs/api/customer_bank_accounts/verify

**Review by LinusU — APPROVED** [review]
Great work! 👍

**Review by galtalmor — APPROVED** [review]
Nice! 👍

**Comment by PranavSenthilnathan** [comment]
Congratulations on your first DefinitelyTyped contribution! Thank you for being a part of the community!

**Pull request merged**